conversation_reconstruction_dataflow/ingest_revisions/dataflow_main.py

```diff
@@ -72,7 +75,21 @@ def run(known_args, pipeline_args):
   with beam.Pipeline(options=pipeline_options) as p:
     pcoll = (p | ReadFromText(known_args.input)
                    | beam.ParDo(WriteDecompressedFile())
-                   | beam.io.Write(bigquery.BigQuerySink(known_args.table, schema=known_args.schema, validate = True)))
+                   | beam.Map(lambda x: ('{week}at{year}'.format(week=x['week'], year=x['year']), x))
+                   | beam.GroupByKey()
+                   | beam.ParDo(WriteToStorage()))
+
+class WriteToStorage(beam.DoFn):
+  def process(self, element):
+      (key, val) = element
+      week, year = [int(x) for x in key.split('at')]
+      path = known_args.output + 'date-{week}at{year}/revisions.json'.format(week=week, year=year)
```

Comment:
Should the output path end in * so that it gets further sharded (making more parallel reading possible...) ?

Reply:
I tried path ends with * and it seems like the code takes the meaning literally, it creates a file named 'revisions*.json'.

Reply:
Look into this; you definitely can write sharded files. Maybe in python API you have an argument to specify, but you definitely want sharded output here.

Reply:
We could do manual sharding, when the number of lines go to a certain size, we manually shard it.

Reply:
We 'could' but lets not :)  Sounds like you're saying you need some help figuring out how to write sharded files... I'll have a look later today/tomorrow.

Reply:
Yes, thanks for the help. Also could you look at this: https://console.cloud.google.com/dataflow/jobsDetail/locations/us-central1/jobs/2018-02-22_08_49_43-5186810640516118154?project=wikidetox-viz

It seems like the ParDo(Write) will not be started until all the GroupBy is finished, sharding might solve this though, I'm not sure.

Reply:
Exactly; if you write all to one file, you'll create a bad synchronization point. You want to be writing to a sharded fileset to allow more and better parallelism, and avoid potential memory explosions too.

Reply:
Great then we found the cause to that problem at least.

So the specific problem here is in the ParDo(Write) function where I used filesystems.create() to create a write channel but was not able to shard from there.

Reply:
The stackoverflow question here:

https://stackoverflow.com/questions/48956426/python-automatic-sharding-in-filesystems-create-for-a-writing-channel-in-data

Reply:
Doing I/O inside of a DoFn is frowned upon, you should be using proper Sinks that know how to shard for you. You are trying to Partition by month and year? There's many ways to do that
https://beam.apache.org/documentation/sdks/pydoc/2.4.0/apache_beam.transforms.core.html?highlight=partition#apache_beam.transforms.core.Partition

For whatever solution you choose, you should Emit() json lines as early as possible and then treat everything after that as line oriented string files which Dataflow knows how to handle well.

Reply:
I think I agree with Jeff, I think you want to be emitting your data, and then writing a custom Sink for your sharding to get custom output locations like this. And probably within that custom output path, using standard text file sharding for your jsonlines final output. That will also be must faster I think.

Reply:
No custom sink needed - you can Partition() by year and Partition() by month and then use the provided Sinks. Or you can just Partition() by month and year directly. Writing sinks in dataflow is very hard.

Reply:
Thanks for the suggestion, I've put this up as an issue, will fix soon.